src/shogun/optimization/FirstOrderSAGCostFunctionInterface.h

```diff
@@ -0,0 +1,161 @@
+/*
+ * This software is distributed under BSD 3-clause license (see LICENSE file).
+ *
+ * Authors: Elfarouk
+ */
+
+#ifndef FIRSTORDERSAGCOSTFUNCTIONINTERFACE_H
+#define FIRSTORDERSAGCOSTFUNCTIONINTERFACE_H
+
+#include <stan/math.hpp>
+#include <shogun/mathematics/eigen3.h>
+#include <shogun/lib/config.h>
+#include <shogun/optimization/FirstOrderSAGCostFunction.h>
+#include <functional>
+#include <vector>
+#include <shogun/lib/SGMatrix.h>
+#include <shogun/lib/SGVector.h>
+
+using Eigen::Dynamic;
+using Eigen::Matrix;
+using stan::math::var;
+using std::function;
+
+namespace shogun
+{
+	/** @brief The first order stochastic cost function base class for
+	 * implementing
+	 *  The SAG Cost function
+	 *
+	 * The class gives the implementation used in first order stochastic
+	 * minimizers
+	 *
+	 * The cost function must be Written as a finite sample-specific sum of
+	 * cost.
+	 * For example, least squares cost function,
+	 * \f[
+	 * f(w)=\frac{ \sum_i{ (y_i-w^T x_i)^2 } }{2}
+	 * \f]
+	 * where \f$(y_i,x_i)\f$ is the i-th sample,
+	 * \f$y_i\f$ is the label and \f$x_i\f$ is the features
+	 */
+	class FirstOrderSAGCostFunctionInterface : public FirstOrderSAGCostFunction
+	{
+	public:
+		/** Consturctor that initializes X, y, trainable_parameters, and the
+		 * cost_function */
+		FirstOrderSAGCostFunctionInterface(
+		    SGMatrix<float64_t>* X, SGMatrix<float64_t>* y,
+		    Matrix<var, Dynamic, 1>* trainable_parameters,
+		    Matrix<function<var(int32_t)>, Dynamic, 1>* cost_for_ith_point,
+		    function<var(Matrix<var, Dynamic, 1>*)>* total_cost);
+
+		/** Default constructor, use setter helpers to set X, y,
+		*trainable_parameters, and
+		*		cost_function
```

Comment:
This is not informative doc :-) I'd remove very general comments like this one, they add noise.